sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ListingFileCatalog.scala

```diff
@@ -58,10 +56,16 @@ class ListingFileCatalog(
   }
 
   override protected def leafFiles: mutable.LinkedHashMap[Path, FileStatus] = {
+    if (cachedLeafFiles eq null) {
+      refresh()
+    }
     cachedLeafFiles
   }
 
   override protected def leafDirToChildrenFiles: Map[Path, Array[FileStatus]] = {
+    if (cachedLeafDirToChildrenFiles eq null) {
+      refresh()
```

Comment:
There is a side effect. `refresh()` rest the `cachedPartitionSpec` to null, which may cleared already inferred partition information. 

```
  override def refresh(): Unit = {
    val files = listLeafFiles(paths)
    cachedLeafFiles =
      new mutable.LinkedHashMap[Path, FileStatus]() ++= files.map(f => f.getPath -> f)
    cachedLeafDirToChildrenFiles = files.toArray.groupBy(_.getPath.getParent)
    cachedPartitionSpec = null
  }
```